File: src/p0020/python/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        def is_match(p, q):
            return p == '(' and q == ')' or \
                   p == '[' and q == ']' or \
                   p == '{' and q == '}'
        stk = []
        for x in s:
            if len(stk) == 0:
                stk.append(x)
            else:
                if is_match(stk[-1], x):
                    stk.pop()
                else:
                    stk.append(x)
        return len(stk) == 0

File: src/p0020/cpp/solution.cpp
class Solution {
public:
    bool isValid(const string &s) {
        stack<char> stk;
        for (auto x : s) {
            if (stk.empty()) {
                stk.push(x);
            } else {
                if (isMatch(stk.top(), x)) {
                    stk.pop();
                } else {
                    stk.push(x);
                }
            }
        }
        return stk.empty();
    }

    bool isMatch(char p, char q) {
        return (p == '(' && q == ')') ||
               (p == '{' && q == '}') ||
               (p == '[' && q == ']');
    }
};
